api/v1/testclustertopology_types.go

```diff
@@ -166,23 +177,19 @@ type TiDBCluster struct {
 }
 
 func (c *TiDBCluster) AllHosts() map[string]struct{} {
+	components := []string{TiDBField, TiKVField, PDField, MonitorField, GrafanaField}
 	result := map[string]struct{}{
 		c.Control: {},
 	}
-	for _, item := range c.TiDB {
-		result[item.Host] = struct{}{}
-	}
-	for _, item := range c.TiKV {
-		result[item.Host] = struct{}{}
-	}
-	for _, item := range c.PD {
-		result[item.Host] = struct{}{}
-	}
-	for _, item := range c.Monitor {
-		result[item.Host] = struct{}{}
-	}
-	for _, item := range c.Grafana {
-		result[item.Host] = struct{}{}
+
+	val := reflect.ValueOf(*c)
```

Comment:
At least we should avoid reflecting fields. We can define an interface `ClusterServer` and implement it for all servers.
```go
type ClusterServer interface {
	Host() string
}

func (t *TiDBSpec) Host() string {
	return t.Host
}

var _ ClusterServer = &TiDBSpec{}

...
```

Then rewrite code following:

```go
for i := 0; i < val.Type().NumField(); i++ {
	  if checkIn(components, val.Type().Field(i).Name) {
		    field := val.Field(i)
		    for j := 0; j < field.Len(); j++ {
			    result[field.Index(j).Interface().(ClusterServer).Host()] = struct{}{}
		    }
	  }
}
```

In this way, the compiler will refuse to compile code after a field or method renaming.

Reply:
When we alter fields of cluster components, we can changed components slice.

Reply:
It's a good idea to add an interface to check.

Reply:
> Type '*TiDBSpec' has both field and method named 'Host'
Because we have exposed a host field, we can't expose another host method. Host is a necessary property of all components. Is it necessary to add interface?